### ISSUE TITLE
toString() GameObject update

### ISSUE THREAD

**Issue by SolarLune** [GoranM/bdx#151]
Hey. toString() for GameObject's a bit lengthy at the moment; I think it could be distilled down to a simpler listing using just the GameObject name. I think most times, you don't really care about the class name it extends, or its ID, but rather just what object it is.

**Comment by GoranM** [comment]
The class may be relevant in the future, if/when we provide the ability to specify an arbitrary class for a given object, instead of just using the one with the same name.

The ID can be quite valuable: You can quickly determine if it's a list of different game objects, or a list of references to a single game object in memory.

**Comment by SolarLune** [comment]
Even if we provided the ability to specify an arbitrary class for a given object, I think that having the class names in whole is pretty awkward. I can't see it being an issue, or even possible for the developer to have multiple objects of the same name with different scripts. The developer would know which class each object is using. If you want to keep it, trimming the class name down to the actual class name instead of the whole class-path might be advised.

While the ID is valuable in actual development, I wouldn't think the need to differentiate between references to different objects or references to a single object would come up often. If it does come up, the developer would know which is which. Even if the ID is very useful for this purpose, it's awkward to read a string of five or six letters and numbers and check them, going back and forth for each pair. 

If the ID is kept, I'd recommend a game object ID variable that would be kept for this purpose. Each game object instantiated would get a unique ID, starting with 1. It could be hexadecimal to allow for a maximum of values per place to limit large numbers (so you wouldn't really have IDs ranging into the hundreds or thousands). 

Basically, I think I'd rather have this:

`[Cube, Player, House]`

over this:

`[Cube:GameObject:1, Player:Player:1B, House:Building:1F]`

And I think I'd rather have that than this:

`[Cube:com.nilunder.bdx.GameObject:$19F67A, Player:...`

or however it is currently.

**Comment by GoranM** [comment]
> The developer would know which class each object is using.

The developer could think they know, but the class they actually specified could be different.

> If it does come up, the developer would know which is which.

How?

> Even if the ID is very useful for this purpose, it's awkward to read a string of five or six letters and numbers and check them, going back and forth for each pair.

It doesn't seem awkward to me. It's just an identifier hash, and in the scenario I described, one doesn't need to go "back and forth" - if they're the same, that will stick out.

In cases where deeper scrutiny is required, it's probably not something you want to filter "by eye" in the first place - even with decimal identifiers (assuming you find those easier to differentiate), you could easily overlook something.

Overall, I don't think it's a good idea to override default toString to provide less information; that would surprise many java programmers who expect to see the class and the hash.

If you want/need names, you can just filter down to that - Everyone should have their own utilities to provide information based on their needs/preferences.

**Comment by SolarLune** [comment]
> The developer could think they know, but the class they actually specified could be different.

OK? We're not concerned with making the printing process proof against mistakes; we're concerned with making it streamlined and easy to use. If the developer makes a mistake, they made a mistake. Printing out the object should not really be the means by which a developer confirms that it's been assigned the correct class. That's what the new panel in Blender will do.

> How?

They coded it. They can use code to find out if references are equal - there's not a huge reason to do it via printing out the objects. Like I said, if you want to keep it, I could see the benefit; I'd rather them just be shorter.

> It doesn't seem awkward to me. It's just an identifier hash, and in the scenario I described, one doesn't need to go "back and forth" - if they're the same, that will stick out.

I don't think so at all. It seems like a jumble of numbers and letters six characters long won't stick out, especially considering the length of the "info string" for each object. Each string can occupy almost an entire line in the console; it seems like it would be extremely easy to read the list for even just a few objects and end up missing duplicates.

> In cases where deeper scrutiny is required, it's probably not something you want to filter "by eye" in the first place - even with decimal identifiers (assuming you find those easier to differentiate), you could easily overlook something.

So then it doesn't matter, right? Decimal or hashcode, you're going to go with a code-based solution to confirm you don't have duplicates in a collection. If this is the case, then it doesn't really matter too much to print out the hashCode.

> Overall, I don't think it's a good idea to override default toString to provide less information; that would surprise many java programmers who expect to see the class and the hash.

That's something to consider, but it's also worth considering that this isn't a traditional Java engine. It's a bit unnecessary to print out the class of an object that you assigned yourself, and that you can see yourself in a GUI. If they didn't assign the class themselves (and the new panel does exist), then the class should be the same name as the object, which is redundant. If/when the panel's implemented, you'll be able to select the PlayerCollision object and will be able to see that it is using the Player class, for example. It's not really necessary to print that out.

> If you want/need names, you can just filter down to that - Everyone should have their own utilities to provide information based on their needs/preferences.

That's true, and I agree with you. By that logic, however, you could just as easily do the opposite - provide a basic toString() function that prints less information, and allow the developer override it to print out more information - the object name, Java class, hashCode, and so forth.

**Comment by GoranM** [comment]
> Printing out the object should not really be the means by which a developer confirms that it's been assigned the correct class. That's what the new panel in Blender will do.

The fact that the information is displayed somewhere in blender doesn't help much if you don't know where to look (on which object).

> it seems like it would be extremely easy to read the list for even just a few objects and end up missing duplicates

If they all have the same id, that's quite obvious, in my experience.

> So then it doesn't matter, right? Decimal or hashcode, you're going to go with a code-based solution to confirm you don't have duplicates in a collection. If this is the case, then it doesn't really matter too much to print out the hashCode.

It depends on what I'm looking for.

If I want to confirm that I don't just have a list of multiple references to the same object in memory, printing out the list will show that, and there would be no need to create my own "code-based solution".

It also depends on the size of the list; if it's just a few objects, printing the list would probably suffice, even for something more specific.

> That's true, and I agree with you. By that logic, however, you could just as easily do the opposite - provide a basic toString() function that prints less information, and allow the developer override it to print out more information

Yes, but why should the default be less information?

> this isn't a traditional Java engine

It's not about the engine: We're talking about the defaults of the Java language. By default, when you print a list, you get [class@id, ...] - Java programmers would expect that, I think.

But, ok, if you just want names, I'll merge that PR - I don't really care that much one way or another, I just wanted to make it clear that you're cutting out potentially useful information, and that it probably goes against the expectations of most Java programmers.

**Comment by SolarLune** [comment]
> The fact that the information is displayed somewhere in blender doesn't help much if you don't know where to look (on which object).

I don't know what you mean; how would the developer not know where to look? They know what objects are in the list. They assigned the class to the object in Blender. They can check those objects in Blender and know what class they're using.

> If they all have the same id, that's quite obvious, in my experience.

I dunno, I can't see having to remember six character IDs and scroll through a console window to compare them obvious or easy. It seems like anything but.

> If I want to confirm that I don't just have a list of multiple references to the same object in memory, printing out the list will show that, and there would be no need to create my own "code-based solution".

> It also depends on the size of the list; if it's just a few objects, printing the list would probably suffice, even for something more specific.

You just said that if it needed deeper scrutiny, it's not something that you would likely filter "by eye".

> Yes, but why should the default be less information?

Because I personally find it difficult to read.

But it's fine; I'm just one guy, and if my opinion is unpopular, then it shouldn't get merged in. This would be a prime time for actual Java users to give their opinion, or even just others who develop for or have used the engine. Maybe I'll pose a question on StackOverflow to get their opinions.

**Comment by GoranM** [comment]
> I don't know what you mean; how would the developer not know where to look?

I guess this probably never happens to you, but on occasion, I will be convinced that I did something (like set/change a specific class for an object), even though I didn't, and the resulting bugs are not obvious enough to make me think otherwise (not as soon as I would if I could just print the info, and notice a mismatch).

Beyond that, something could go wrong in the exporter, or on the engine side, where objects are mapped incorrectly; being able to just scan a list print out, and notice the discrepancy, would be immediately helpful.

Actually, there was a bug with instantiators, early in the development of BDX, which I managed to quickly localize, because I could just print the objects, and see that they're all GameObject - It wasn't something that required deep scrutiny, because I know that most objects should be some other class, but everywhere my eyes scan, they see GameObject.

> I dunno, I can't see having to remember six character IDs and scroll through a console window to compare them obvious or easy. It seems like anything but.

You wouldn't have to check every hash against every other hash: If all the references in the list are supposed to point to unique objects in memory, but wherever your eyes scan, they see the same hash, the nature of the problem is obvious.

You wouldn't waste time actually comparing all the hashes, of course.

> You just said that if it needed deeper scrutiny, it's not something that you would likely filter "by eye".

The cases I listed don't require deeper scrutiny.

> Maybe I'll pose a question on StackOverflow to get their opinions.

I'm not sure that a StackOverflow thread would reliably reflect the preferences of most Java programmers. We can know for sure that Java programmers would expect to see the class and hash, because that's what toString provides by default.

Whether they would prefer just names, in BDX context, is something we can wait to find out, as we (slowly) ramp up the user base.

For now, if you want just names, I'm ok with that, as I said.
